crates/lib3h_protocol/src/uri.rs

```diff
@@ -0,0 +1,188 @@
+use crate::{error::Lib3hProtocolError, Address};
+use std::convert::TryFrom;
+use url::Url;
+
+//--------------------------------------------------------------------------------------------------
+// UriScheme
+//--------------------------------------------------------------------------------------------------
+
+static AGENT_SCHEME: &'static str = "agentid";
+static TRANSPORT_SCHEME: &'static str = "transportid";
+static MEMORY_SCHEME: &'static str = "mem";
+static UNDEFINED_SCHEME: &'static str = "none";
+
+pub enum UriScheme {
+    Agent,
+    Transport,
+    Memory,
+    Undefined,
+    Other(String),
+}
+
+impl From<UriScheme> for &str {
+    fn from(scheme: UriScheme) -> &'static str {
+        match scheme {
+            UriScheme::Agent => AGENT_SCHEME,
+            UriScheme::Transport => TRANSPORT_SCHEME,
+            UriScheme::Memory => MEMORY_SCHEME,
+            UriScheme::Undefined => UNDEFINED_SCHEME,
+            UriScheme::Other(_) => "",
+        }
+    }
+}
+
+impl From<UriScheme> for String {
+    fn from(scheme: UriScheme) -> String {
+        match scheme {
+            UriScheme::Agent => AGENT_SCHEME.into(),
+            UriScheme::Transport => TRANSPORT_SCHEME.into(),
+            UriScheme::Memory => MEMORY_SCHEME.into(),
+            UriScheme::Undefined => UNDEFINED_SCHEME.into(),
+            UriScheme::Other(s) => s.clone(),
+        }
+    }
+}
+
+//--------------------------------------------------------------------------------------------------
+// Lib3hUri
+//--------------------------------------------------------------------------------------------------
+
+#[derive(Shrinkwrap, Debug, Clone, PartialEq, Eq, PartialOrd, Ord, Hash, Serialize, Deserialize)]
+#[shrinkwrap(mutable)]
+pub struct Lib3hUri(pub Url);
+
+impl Lib3hUri {
+    pub fn is_scheme(&self, scheme: UriScheme) -> bool {
+        let s: String = scheme.into();
+        self.scheme() == s
+    }
+    pub fn with_transport_id(transport_id: &Address, agent_id: &Address) -> Self {
+        let url = Url::parse(&format!(
+            "{}:{}?a={}",
+            TRANSPORT_SCHEME, transport_id, agent_id
+        ))
+        .unwrap();
+        Lib3hUri(url)
+    }
+    pub fn with_agent_id(agent_id: &Address) -> Self {
+        let url = Url::parse(&format!("{}:{}", AGENT_SCHEME, agent_id)).unwrap();
+        Lib3hUri(url)
+    }
+    pub fn with_undefined(other: &str) -> Self {
```

Comment:
What I had intended with this was to create a "NULL" URI, i.e. what that was literally like `None`  which has the semantic meaning of we don't have a URI here, not just one where the scheme is the undefined scheme.  

I see you made the "Fixme's" into Undefined's and I don't think that's right.  They should be FIXME's within in one of the scheme's that they are supposed to be.